[Tianran] iP

Nature711 opened the pull request:
# DinoPro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DinoPro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/Nature711/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---
If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```ruby
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Nature711 committed:
add greet, echo, exit

Nature711 committed:
add add, list

Nature711 committed:
add mark as done feature

Nature711 committed:
updated implementation of mark as done feature

Nature711 committed:
add todo, deadline, event

Nature711 committed:
add automated text ui testing

Nature711 committed:
add customized exceptions

Nature711 committed:
add delete feature

Nature711 committed:
fixed bugs

Nature711 committed:
fix bugs & update ui testing

Nature711 committed:
add save feature

Nature711 committed:
add date and times format

Nature711 committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Event.java

Nature711 committed:
fix bugs in date time formatter

Nature711 committed:
add exception handling for date display

Nature711 committed:
add Ui, Storage, Parser, TaskList

Nature711 committed:
group classes into packages

Nature711 committed:
add Junit tests

Nature711 committed:
create JAR file

Nature711 committed:
add find feature

Nature711 committed:
add javadoc

Nature711 committed:
fix coding standard issues

Nature711 committed:
Revert "fix coding standard issues"

This reverts commit 55469e7ce2e835d22873761c55ee63eedf0ad011.

Nature711 committed:
Revert "Revert "fix coding standard issues""

This reverts commit bb3c42317141b2bfcaa2a8dabecaf6fa79c3344a.

Nature711 committed:
Merge branch 'branch-A-CodingStandard'

Nature711 committed:
make use of enum class in Parser class

Nature711 committed:
fix coding standard issues

Nature711 committed:
fix minor issues

Nature711 committed:
Merge branch 'add-gradle-support' of https://github.com/Nature711/ip into add-gradle-support

Nature711 committed:
add gradle support to project

Nature711 committed:
Merge branch 'add-gradle-support'

Nature711 committed:
add checkstyle plugin

Nature711 committed:
fix checkstyle issues

Nature711 committed:
Merge branch 'add-checkstyle'

Nature711 committed:
add GUI to java application

Nature711 committed:
Merge branch 'branch-Level-10'

calvintanwj commented on src/main/java/dino/Dino.java:
```diff
@@ -0,0 +1,81 @@
+ * It helps the user to keep track of various things
+ */
+public class Dino {
+    private final TaskList taskList;
```
Why not call it "tasks" instead of "taskList"? :)

calvintanwj commented on src/main/java/dino/Dino.java:
```diff
@@ -0,0 +1,81 @@
+     * @return the output message after execution
+     * @throws DinoException the general set of exceptions
+     */
+    public String processCommand(String command) throws DinoException {
+        Parser.CMDTYPE cmdType = Parser.parse(command);
+        switch (cmdType) {
+        case TODO:
+        case EVENT:
+        case DEADLINE: {
+            Command cmd = new AddTaskCommand(command, cmdType);
+            return cmd.execute(storage, taskList);
+        }
+        case DONE:
+        case DELETE: {
+            Command cmd = new MarkCommand(command, cmdType);
+            return cmd.execute(storage, taskList);
+        }
+        case LIST: {
+            Command cmd = new ListCommand();
+            return cmd.execute(storage, taskList);
+        }
+        case FIND: {
+            Command cmd = new FindCommand(command);
+            return cmd.execute(storage, taskList);
+        }
+        case BYE: {
+            return ui.exit(storage, taskList);
+        }
+        default: {
+            throw new InvalidInputException();
+        }
+        }
```
Very nice! Easy to understand and readable.

calvintanwj commented on src/main/java/dino/Dino.java:
```diff
@@ -0,0 +1,81 @@
+package dino;
+
+import dino.command.*;
+import dino.exception.*;
```
Perhaps it would be better to explicitly name your imports for better documentation.

Moley456 replied:
Would it be better if your imported classes are explicitly stated for better documentation? This would also comply with the coding standards in the CS2103T website as well 😄

calvintanwj commented on src/main/java/dino/command/AddTaskCommand.java:
```diff
@@ -0,0 +1,123 @@
+     * @param taskString the user input command
+     * @param taskType the type of the task, which can only be one of the type ToDo, Deadline, or Event
+     */
+    public AddTaskCommand(String taskString, Parser.CMDTYPE taskType) {
```
hm, this is interesting. May I know why you decided to group the adding of tasks as one command, instead of separating them such as "AddListCommand", "AddDeadlineCommand"? ... I think as the code-base gets bigger, and if you do choose to implement more different types of tasks, this class and in particular the execute and getTaskDescription methods can get pretty messy.

calvintanwj commented on src/main/java/dino/command/AddTaskCommand.java:
```diff
@@ -0,0 +1,123 @@
+     * @throws EmptyTaskDescriptionException if the description of the task is empty
+     * @throws TimeNotSpecifiedException if the task is of type Deadline or Event but the time is not specified
+     */
+    public static String getTaskDescription(String s, Parser.CMDTYPE type) throws EmptyTaskDescriptionException, TimeNotSpecifiedException {
```
Perhaps the parameter "s" can be named better, may I suggest something like "userInput" or "inputString"

Moley456 replied:
Would it be better if the parameter "s" was named "userCommand" or "userInput"? This would allow others understand the code more easily.

calvintanwj commented on src/main/java/dino/command/AddTaskCommand.java:
```diff
@@ -0,0 +1,123 @@
+                throw new TimeNotSpecifiedException(type.toString());
+            }
+        }
+        default: {
+            break;
+        }
+        }
+        return s;
```
Why not shift the "return s;" into the default branch. It seems more readable to me. :) Additionally for switch statements, I believe you can remove the curly braces for each case.
```suggestion
        default: 
            return s;
            break;
        }

```

Moley456 replied:
Hmm do you think it's better to just move the return into the default case and remove the break statement? I feel that it would make the code look cleaner! Also, the curly braces could be removed as well in each of the cases.

```suggestion
        default: 
            return s;
        }
```

calvintanwj commented on src/main/java/dino/command/AddTaskCommand.java:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    /**
+     *
```
There seems to be a missing description here 😄

calvintanwj commented on src/main/java/dino/command/Command.java:
```diff
@@ -0,0 +1,15 @@
+ */
+public abstract class Command {
+
+    abstract public String execute(Storage storage, TaskList taskList) throws InvalidFormatException, TimeNotSpecifiedException, EmptyTaskDescriptionException, InvalidIndexException, TaskAlreadyDoneException, InvalidInputException, IndexNotSpecifiedException, EmptyListException, TaskNotFoundException;
```
Perhaps you can have a parent class like DinoException, where all these exceptions extend from this DinoException class. Then the method signature could be throws DinoException instead, of having such a long method signature. I think it would be quite a headache, when you add more exceptions and will have to change the method signatures of your command classes.

calvintanwj replied:
After reading further down, I realize you do have a DinoException, so why not just throws DinoException instead? :)

calvintanwj commented on src/main/java/dino/command/MarkCommand.java:
```diff
@@ -0,0 +1,91 @@
+            InvalidInputException, IndexNotSpecifiedException {
+        int index = getIndex(taskString, markType);
+        switch (markType) {
+        case DONE: {
+            return taskList.markTaskDone(index);
+        }
+        case DELETE: {
+            return taskList.deleteTask(index);
+        }
+        default: {
+            throw new InvalidInputException();
+        }
```
Same as before. I don't think there is a need for the curly braces.
```suggestion
        case DONE: 
            return taskList.markTaskDone(index);
        case DELETE: 
            return taskList.deleteTask(index);
        default: 
            throw new InvalidInputException();
```

calvintanwj commented on src/main/java/dino/task/TaskList.java:
```diff
@@ -0,0 +1,117 @@
+     * @throws EmptyListException if the current task list is empty
+     */
+    public String printTaskList() throws EmptyListException {
+        if (taskList.isEmpty()) throw new EmptyListException();
```
Don't forget to adhere to egyptian brackets. This is readable to me, but best to adhere to coding standards :)
```suggestion
        if (taskList.isEmpty()) { 
            throw new EmptyListException();
        }
```

Moley456 replied:
According to the coding standards, I think there should be Egyptian brackets here instead of having everything on one line.

calvintanwj commented on src/main/java/dino/task/TaskList.java:
```diff
@@ -0,0 +1,117 @@
+     */
+    public String printTaskList() throws EmptyListException {
+        if (taskList.isEmpty()) throw new EmptyListException();
+        String list = "";
+        for (int i = 0; i < taskList.size(); i++) {
+            list += (i + 1) + ". " + taskList.get(i) + "\n";
+        }
```
May I suggest using a StringBuilder instead, since String is immutable in Java, and concatenation creates a new String each time, resulting in O(n^2) time complexity. I think for simple string concatenation, using the '+' operator is fine, but because you are looping through all the tasks in your taskList, I think it can get quite costly.

calvintanwj commented on src/main/java/dino/util/Storage.java:
```diff
@@ -0,0 +1,132 @@
+import java.util.List;
+import java.util.Scanner;
+
+import dino.exception.*;
+import dino.task.*;
```
Once again, I think it's better to use explicit import statements. I believe for IntelliJ, it automatically uses wildcard imports, if you have 3 or more imports from the same package. This limit can be changed or toggled off within your IDE settings.

calvintanwj reviewed (commented):
Overall, looks really good 👍  A few cosmetic issues here and there such as the curly braces in switch statements, and your import statements. Naming conventions for certain parameters can be improved as well (instead of using "s" and "t"). Other than that, the code is readable and structured really well.

Moley456 commented on src/main/java/dino/exception/DinoException.java:
```diff
@@ -0,0 +1,30 @@
+        super(message);
+    }
+}
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
```
You could probably remove the empty lines here 😅

Moley456 commented on src/main/java/dino/command/MarkCommand.java:
```diff
@@ -0,0 +1,91 @@
+
+
+    /**
+     * Extracts the index indicated by the done or delete command
```
Small detail here but I think your description and parameter section should have an empty line separating them, according to the coding standards.

Moley456 reviewed (commented):
Everything looks good. Clean code with just some minor changes to be made to adhere to the coding standards. Keep it up! 👍🏻

Nature711 committed:
add vararg feature in find command to support searching of multiple keywords

Nature711 committed:
Add assertions to explictily state the assumptions made during coding

This will allow testing the correctness of any assumptions that have been made in the program.

Mainly concerns with assertions that the input is not null before further processing it, preventing any unexpected errors from occuring

Nature711 committed:
Add assertions to explictily state the assumptions made during coding

This will allow testing the correctness of any assumptions that have been made in the program.

Mainly concerns with assertions that the input is not null before further processing it, preventing any unexpected errors from occuring

Nature711 committed:
Merge pull request #1 from Nature711/A-Assertions

A assertions

Nature711 committed:
Improve code quality according to the lecture topic

Critically examines the code and refactor to improve the code quality at necessary places, improving readability of code

Changes include avoiding complicated expressions by separating into intermediate steps, changing the method implementation to make the functionality of the code more explicit, as well as getting rid of redundant initializations of local variables.

Nature711 committed:
Merge pull request #2 from Nature711/A-CodingQuality

Improve code quality according to the lecture topic

Nature711 committed:
Make use of stream function to implement the find feature

Nature711 committed:
Merge pull request #3 from Nature711/heads/A-Streams

Make use of stream function to implement the find feature

Nature711 committed:
no message

Nature711 committed:
no message

Nature711 committed:
Merge pull request #4 from Nature711/branch-A-Streams

new branch

Nature711 committed:
Add yml file to project

Nature711 committed:
Merge branch 'master' of https://github.com/Nature711/ip

Nature711 committed:
Add edit task feature to support the updating of an existing item without having to delete it first

Nature711 committed:
Merge branch 'branch-Extension-Update'

Nature711 committed:
Improve GUI design

Nature711 committed:
Fix issues with parsing inputs

Nature711 committed:
Add screenshot for Ui

Nature711 committed:
Add User Guide

Nature711 committed:
Edit User Guide

Nature711 committed:
Edit User Guide again

Nature711 committed:
Add necessary dependencies to the build.gradle to pack JavaFX libraries into the generated JAR file

Nature711 committed:
Attempt to fix JAR file issue again

Nature711 committed:
Modify dependencies in build.gradle file

Nature711 committed:
Merge remote-tracking branch 'origin/branch-A-Assertions'

# Conflicts:
#	src/main/java/dino/task/TaskList.java
